src/core/packages/saved-objects/base-server-internal/src/utils/pick_values_based_on_structure.test.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { fc } from '@fast-check/jest';
+import { pickValuesBasedOnStructure, getFlattenedKeys } from './pick_values_based_on_structure';
+
+describe('getFlattenedKeys', () => {
+  test('empty object', () => {
+    expect(getFlattenedKeys({}).sort()).toEqual([]);
+  });
+
+  test('simple object', () => {
+    expect(getFlattenedKeys({ a: 1, b: 2, c: 3 }).sort()).toEqual(['["a"]', '["b"]', '["c"]']);
+  });
+
+  test('simple object 2', () => {
+    const obj = {
+      a: {},
+    };
+    expect(getFlattenedKeys(obj).sort()).toEqual(['["a"]']);
+  });
+
+  test('simple object 3', () => {
+    const obj = {
+      a: [[]],
+    };
+    expect(getFlattenedKeys(obj).sort()).toEqual(['["a"][0]']);
+  });
+
+  test('complex object 1', () => {
+    const obj = {
+      a: 1,
+      b: { c: 2 },
+      d: [{ e: 1 }, { e: 2 }],
+    };
+
+    expect(getFlattenedKeys(obj).sort()).toEqual([
+      '["a"]',
+      '["b"]["c"]',
+      '["d"][0]["e"]',
+      '["d"][1]["e"]',
+    ]);
+  });
+});
+
+describe('pickValuesBasedOnStructure', () => {
+  test('picks the values of the target', () => {
+    const a = { v1: 'a', v2: [{ c: 1, d: 1 }] };
+    const b = { v1: 'b', v2: [{ c: 2, d: 2 }, { e: 4 }], another: 'value', anArray: [1, 2, 3] };
+
+    expect(pickValuesBasedOnStructure(a, b)).toEqual({ v1: 'b', v2: [{ c: 2, d: 2 }] });
+  });
+  test('picks a subset of values from target', () => {
+    /**
+     * The `keys` `Arbitrary` represents words with possible numbers like `loremv123`
+     * that might be used in saved objects as key names.
+     */
+    const keys = fc
+      .array(
+        fc.tuple(
+          fc.lorem({ mode: 'words', maxCount: 1 }),
+          fc.option(fc.nat({ max: 1_000 })),
+          fc.option(fc.constant('.'))
+        ),
+        {
+          minLength: 10,
```

Comment:
When the source contains an empty array the target structure array values are returned in the picked result. I think if we allow a `minLength: 0` the test might pick up this scenario
```
const keySource = {
      emptyArr: [],
      emptyObj: {},
      nonEmptyArr: [1],
      nonEmptyObj: { a: 1 },
    };

    const target = {
      emptyArr: [10, 20],
      emptyObj: { a: 10 },
      nonEmptyArr: [10, 20],
      nonEmptyObj: { a: 10, b: 20 },
    };

    const result = pickValuesBasedOnStructure(keySource, target);

    // This should preserve empty arrays and objects
    expect(result).toEqual({
      emptyArr: [],
      emptyObj: {},
      nonEmptyArr: [10],
      nonEmptyObj: { a: 10 },
    }); 
 ```

Reply:
Great catch, this sent me down a slight road with property based testing 😅 . The reason it was not covering this case it bc of how I was merging `objA` and `objB` a bit too naively using `Object.assign({} ...)`. This made the extraction a bit too straightforward and meant the objects were not truly merged as they would be in more realistic scenarios like what you identified. To remedy this I did the following:

1. Update the merging logic to use `deepMerge` with special array merge handling to make sure our test "target" object was a bit closer to what we would see in the wild

2. Added a traditional example test case to show that we are now handling this case correctly...

Let me know what you think!